onnxruntime/core/session/inference_session.h

```diff
@@ -315,9 +377,16 @@ class InferenceSession {
   // The file path of where the model was loaded. e.g. /tmp/test_squeezenet/model.onnx
   std::basic_string<ORTCHAR_T> model_location_;
 
+  // Immutable state for each op in the model. Shared by all executors.
+  // It has a dependency on execution_providers_.
+  std::unique_ptr<SessionState> session_state_;
```

Comment:
Can we avoid converting these member variables to unique_ptr?  Because for each of them we'll need one more heap allocation.

Reply:
We discussed this. Unfortunately, it doesn't seem like it is avoidable. `session_state_` needs session options to be finalized before it can be constructed and this introduces the need to use the smart pointer...